API: deprecate compat and selected lib utils

mtsokol opened the pull request:
Hi @rgommers,

Here I share a WIP PR with a deprecation of a few positions mentioned in #23537 for the purpose of cleaning up main namespace API.

Here I deprecate `compat` module raising a warn on import, and deprecate some functions from `lib` (listed in #12385). Functions are deprecated in such way: When they are imported from the main namespace a warn is printed, when they are called an exception is raised, saying that these methods are expired and they can be used directly from `numpy.lib` (the financial functions are deprecated this way).   
When these functions are imported directly from `numpy.lib` or `numpy.lib.utils` and called a warn is raised.

mtsokol committed:
API: deprecate compat and selected lib utils

rgommers commented on numpy/__init__.py:
```diff
@@ -155,6 +155,15 @@
     from . import matrixlib as _mat
     from .matrixlib import *
 
+    # Deprecations introduced in 1.25.0, 2023-05-xx TODO update day
```
This should now be 1.26.0, since the `maintenance/1.25.x` branch just got created. The plan is to do a 1.26.0 release in September or so, with a number of deprecations and with the build system changes needed for Python 3.12 support.

mtsokol replied:
Sure, changed!

rgommers commented on numpy/__init__.py:
```diff
@@ -155,6 +155,15 @@
     from .matrixlib import *
 
+    # Deprecations introduced in 1.25.0, 2023-05-xx TODO update day
+    from . import compat
```
This will generate a warning. Hence I think you don't want to do this import (`import numpy` should be warning-free), but leave it to `__getattr__` which already can make it available as a lazy import (which will then emit the warning from `compat/__init__.py` on access).

mtsokol replied:
Sure! I moved `compat` import to `__getattr__` method.

rgommers commented on pytest.ini:
```diff
@@ -13,6 +13,7 @@ filterwarnings =
     ignore::UserWarning:cpuinfo,
     ignore: divide by zero encountered in log
     ignore: invalid value encountered in log
+    ignore:`np.compat`, which was used during the Python 2 to 3 transition, is deprecated
```
I think this should be left out. We should no longer be using `np.compat` in the test suite or the code base, beyond the single test to check that a deprecation warning is emitted.

mtsokol replied:
Hi @rgommers,
It's clear to me - once I remove this exclusion tests fail in a few places as `compat` is used across codebase to some extension. 

I've got a question - If a given function (e.g. `asbytes` from `numpy.compat.py3k.py`) is used in several places (here this one is used in `defchararray.py` and `_iotools.py`) should I move this specific one to e.g. `numpy._utils` and import from there? I see that `_utils` module serves purpose of a internal miscellaneous helpers.

rgommers replied:
There may be cases where that is needed, and then `_utils` is fine, however I suspect that for most functions in `numpy.compat` they should instead be replaced by the idiomatic Python 3 way of doing things. That compat code is written the way it is to be compatible with Python 2 and 3 at the same time - and we don't have to worry about py2 anymore.

In this case, the function is:
```python
def asbytes(s):
    if isinstance(s, bytes):
        return s
    return str(s).encode('latin1')
```

In Python 3, it should be clear whether something is a string or bytes. In most places it's probably a string already, in which case `.encode('latin1')` can be used directly.

If it's too fiddly to figure out you can be pragmatic and retain `asbytes`; it doesn't matter too much (moving it somewhere private achieves the main goal), but in most cases it should be straightforward to remove the use of the compat functions completely.

mtsokol replied:
It's all clear to me! 
I added a very slimmed file in `_utils` for functions that are still used in `np.savetxt` for converting from and to bytes (in `write_normal` method). 

I other places, where is was clear, I removed the usage of `compat` items.
For example in `np.savetxt` parameter `fmt` accepts `str` and `List[str]` according to the stub and documentation. I dropped an additional acceptance of bytes (It wasn't mentioned in docs, so I assumed it came from compatibility with Python 2)

rgommers commented on numpy/__init__.py:
```diff
@@ -232,9 +241,20 @@
+        __expired_functions__[name] = (
+            f"{name} is deprecated and removed from the "
+            f"main namespace. If you still want to use it "
+            f"please use it directly from np.lib or np.lib.utils"
```
You want to leave them in the main namespace for now, to avoid breaking backwards compatibility right away.

mtsokol replied:
I added it back - the warning is raised only on calling them.

rgommers reviewed (commented):
Thanks for picking this up @mtsokol! A few comments to tweak the approach.

mtsokol committed:
API: remove np.compat usage from codebase

mtsokol commented:
Hi @rgommers,
I pushed new changes with addressed review comments. In the current version the codebase doesn't use/depend on `np.compat`, so in the future release it can be removed by just dropping the module itself. 

`lib/utils` methods covered in this PR are deprecated, but still available and used (therefore I silenced warnings caused by deprecations in `pytest.ini`)

rgommers reviewed (commented):
This looks great, and is very close to mergeable - only two more comments.

seberg commented on numpy/core/_methods.py:
```diff
@@ -3,17 +3,17 @@
 from numpy.core import umath as um
 from numpy.core.multiarray import asanyarray
 from numpy.core import numerictypes as nt
-from numpy.core import _exceptions
```
Leaving it is OK, the alternative is adding it to `numpy/_pyinstaller/hook-numpy.py`

mtsokol replied:
Ah, I didn't notice that - I moved it back

mtsokol committed:
API: remove selected numpy.lib warning suppressions

mtsokol commented on pytest.ini:
```diff
@@ -27,3 +27,7 @@ filterwarnings =
     ignore:\n\n  `numpy.distutils`:DeprecationWarning
 # Ignore mypy >= 0.971 DeprecationWarnings
     ignore:path is deprecated\. Use files\(\) instead:DeprecationWarning:mypy
+# Ignore selected DeprecationWarnings from numpy.lib
```
Clear! I removed these filters and suppressed deprecation warnings on test level where it was needed.

rgommers reviewed (approved):
LGTM now and all green, in it goes. Thanks again @mtsokol!

rgommers commented:
I'll squash-merged this so it's easy to backport to 1.26.x once that branch gets created. @charris I think this is the first PR that should end up in 1.26.0. Do you have a preference for how to set milestone/labels here to keep track of things for that release?

lesteve commented:
It turns out `np.byte_bounds` is used in joblib and has actually been documented since numpy 1.18 (done in #13847). This PR is causing failures in the scikit-learn CI build where we test against the numpy development version (and turn most warnings into errors). From the linked issues, astropy had seen similar failures.

My understanding is that NEP 52 aims at cleaning up the main namespace. I would have expected that `np.byte_bounds` would be deprecated but not necessarily `np.lib.byte_bounds`.

I see a two main ways forward (there may be others):
- `np.lib.byte_bounds` will be removed eventually, so we need to reimplement `byte_bounds` in joblib if we actually need it. The code of `np.lib.byte_bounds` seems simple enough so it should not be too much work.
- `np.lib.byte_bounds` is going to stay, so it needs to be "undeprecated" in numpy and we need to use it in joblib rather than the deprecated `np.byte_bounds`

As a side-comment, the list of undocumented functions in the main namespace from https://github.com/numpy/numpy/issues/12385 is from November 2018 and is not fully accurate, for example #13847 added documentation for other functions, like `deprecate`, `deprecate_with_doc`, etc ...

rgommers commented:
Adding the forward link: gh-24154 un-deprecates `np.byte_bounds` for now. It seems like we agree that it shouldn't be in the main namespace; it could be useful to find it a new home. We'll need to sort out how `numpy.lib` is structured before that can be done, there's no good namespace for it right now.
